### DISCUSSION TITLE
Can you change the date with a filter?

### DISCUSSION THREAD

**Discussion by andrewpbray** [quarto-dev/quarto-cli#9345] [Q&A] [answered]
### Description

Is it possible to run a filter that will be able to run before Quarto processes `date` meta data into its given date format?

I'm looking at a simple filter that looks like this:

```lua
function Meta(el)
  quarto.log.output(el["date"])
end
```

When run on a blank document with `data: 5/5/24`, the result is:

```bash
Inlines {
  [1] Str {
    clone: function: 0x7fbf52893980
    text: "May"
    walk: function: 0x7fbf52893d90
  }
  [2] Space {
    clone: function: 0x7fbf52895320
    walk: function: 0x7fbf52895710
  }
  [3] Str {
    clone: function: 0x7fbf528969d0
    text: "5,"
    walk: function: 0x7fbf52894fa0
  }
  [4] Space {
    clone: function: 0x7fbf528981e0
    walk: function: 0x7fbf528985e0
  }
  [5] Str {
    clone: function: 0x7fbf528998e0
    text: "2024"
    walk: function: 0x7fbf52899ce0
  }
}
```
So the filter is running after Quarto has processed the date field. I get the same output regardless of when I [run the filter](https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases) at `pre-ast` or `pre-quarto`.

The upshot, as far as I can tell, is that something like this isn't possible:

```lua
function Meta(el)
  el["date"] = "1/1/24"
end
```
Or rather, it is possible, but this change of date happens after the date formatter has run, so it will just display as "1/1/24" in the doc instead of "January 1, 2024".

Is there any way to get such a filter to run before the date formatter or to re-run the date formatter?

Thanks!

**Comment by cderv** [comment]
As you found date is processed in Quarto processing before Pandoc runs lua. So this won't be possible. 

Can you say more about what you want to do exactly ? How you would need to change date ? for what purpose ? 

With more information on the why, we can help more on the how. 

thanks !

**Reply by andrewpbray** [reply]
This is for a [filter](https://github.com/qmd-lab/qmd-specific-meta) that allows you to collect all of your document-specific meta in a single array in your `_quarto.yml`. I'm revising the filter so that it can be embedded in a project extension I've been working on  called [scheduled-docs](https://github.com/qmd-lab/scheduled-docs) that allows you to schedule the drafting of your documents. 

Since that functionality requires that you list the document names and a date on which you'd like to publish them, I'd like to pull from that metadata for the date field in these documents instead of needing to specify them in each separate qmd. I'm working on `date` right now, but in principle I'd like to be able to specify any document-level metadata in `_quarto.yml` under `scheduled-docs` and have it propagate it as a top-level key in the document's meta data.

**Reply by cderv** [reply, marked as answer]
I see.  thanks

There are some metadata you won't be able to modify from a Lua filter. You would need to pass them differently  - like using shared metadata file or passing metadata at command line rendering. 

For this specific `date` usage, won't the date of included the same as the rendering as you schedule your document ? 
special cases `today`, `now` or `last-modified` does not help ? 

Unfortunately, I am thinking only of workarounds this limitation, not real solution. Another one that comes to mind is a pre-render script that modify the date in the YAML of the document. 

Though I wonder if variables could be used with the `date` fields (https://quarto.org/docs/authoring/variables.html#var) - they are the closest to pass values. `meta` or `env` are similar options. 

Not sure it helps... 😅

**Reply by andrewpbray** [reply]
> For this specific date usage, won't the date of included the same as the rendering as you schedule your document ?
special cases `today`, `now` or `last-modified` does not help ?

I don't think these will help. `scheduled-docs` will set to `draft: false` docs with dates that precede the `draft-after` date. So a document with a date that's the same as the `draft-after` date will be correct if the document is set to `date: today` but any document with a preceding dates will get set to that same date even though it might have occurred weeks previously.

> Unfortunately, I am thinking only of workarounds this limitation, not real solution. Another one that comes to mind is a pre-render script that modify the date in the YAML of the document.

Modifying the document itself seems messy - I'd prefer to not tinker with document source if possible. An alternative, though, might be to write a small `_metadata.yml` file with the metadata specific to a doc in the same directory as the doc. The post-render script could clean up all those files.

...on second thought that won't work. I don't want to require that every qmd be in its own dir, so that method could result in one set of metadata going to mulitiple documents, which is what I'm trying to avoid.

> Though I wonder if variables could be used with the date fields (https://quarto.org/docs/authoring/variables.html#var) - they are the closest to pass values. meta or env are similar options.

If you can include `{{< meta something >}}` in document metadata and have it evaluated before the Quarto filters process the date and author fields, we might be in business. I'll give that a try and report back!

ps: @cderv when you have a moment, could you give me your thoughts on #9280 ?

**Reply by andrewpbray** [reply]
It looks like neither the `meta` nor `var` shortcodes will work. If you set `foo.date: "1/1/24"` in project metadata then try:

```yaml
---
date: "{{< meta foo.date >}}"
---
```
the document will show the date as 1/1/24. That's to say, it looks like the Quarto shortcodes run after the date is formatted.

I think I'll abandon this effort for now. Thanks for the help brainstorming!

**Reply by cderv** [reply]
Interesting thanks for sharing. 

I wonder if we could change this.  

@dragonstyle if you happen to pass around, do you think we could make date processing happen later ?

**Reply by cscheid** [reply]
@andrewpbray can you open an issue? I'll take a look at it

**Reply by cderv** [reply]
Thanks @cscheid !

**Reply by andrewpbray** [reply]
When putting together a minimal example, I learned that my `qmd-specific-meta.lua` filter had been running in the aforementioned example, so that was how 1/1/24 was getting into date meta. Without the filter, 
```yaml
date: "{{< meta foo.date >}}"
```
Yields "Invalid date" in the document.

So ... I'm not sure this is worth an issue. For my particular use, the crux of the issue seems to be that either the pandoc filter would need to run be able to run before the Quarto filters or a user could tigger an option to re-run the Quarto filters after the pandoc filters run.

**Reply by cscheid** [reply]
> a user could tigger an option to re-run the Quarto filters after the pandoc filters run.

You're allowed to specify more than one filter to be run at different stages, so I think this should be possible.
